compiler/rustc_trait_selection/src/traits/vtable.rs

```diff
@@ -434,37 +412,23 @@ pub(crate) fn supertrait_vtable_slot<'tcx>(
 
 fn trait_refs_are_compatible<'tcx>(
     tcx: TyCtxt<'tcx>,
-    hr_vtable_principal: ty::PolyExistentialTraitRef<'tcx>,
-    hr_target_principal: ty::PolyExistentialTraitRef<'tcx>,
+    vtable_principal: ty::ExistentialTraitRef<'tcx>,
+    target_principal: ty::ExistentialTraitRef<'tcx>,
 ) -> bool {
-    if hr_vtable_principal.def_id() != hr_target_principal.def_id() {
+    if vtable_principal.def_id != target_principal.def_id {
         return false;
     }
 
     let (infcx, param_env) =
         tcx.infer_ctxt().build_with_typing_env(ty::TypingEnv::fully_monomorphized());
     let ocx = ObligationCtxt::new(&infcx);
-    let hr_source_principal =
-        ocx.normalize(&ObligationCause::dummy(), param_env, hr_vtable_principal);
-    let hr_target_principal =
-        ocx.normalize(&ObligationCause::dummy(), param_env, hr_target_principal);
-    infcx.enter_forall(hr_target_principal, |target_principal| {
-        let source_principal = infcx.instantiate_binder_with_fresh_vars(
-            DUMMY_SP,
-            BoundRegionConversionTime::HigherRankedType,
-            hr_source_principal,
-        );
-        let Ok(()) = ocx.eq_trace(
-            &ObligationCause::dummy(),
-            param_env,
-            ToTrace::to_trace(&ObligationCause::dummy(), hr_target_principal, hr_source_principal),
-            target_principal,
-            source_principal,
-        ) else {
-            return false;
-        };
-        ocx.select_all_or_error().is_empty()
-    })
+    let source_principal = ocx.normalize(&ObligationCause::dummy(), param_env, vtable_principal);
+    let target_principal = ocx.normalize(&ObligationCause::dummy(), param_env, target_principal);
+    let Ok(()) = ocx.eq(&ObligationCause::dummy(), param_env, target_principal, source_principal)
+    else {
+        return false;
+    };
+    ocx.select_all_or_error().is_empty()
```

Comment:
This also needs to do a leak check, otherwise this code will still segfault:

```rust
#![feature(trait_upcasting)]

trait Super<U> {
    fn call(&self) where U: HigherRanked {}
}
impl<T> Super<T> for () {}

trait HigherRanked {}
impl HigherRanked for for<'a> fn(&'a ()) {}

trait Unimplemented {}
impl<T: Unimplemented> HigherRanked for T {}

trait Sub: Super<fn(&'static ())> + Super<for<'a> fn(&'a ())> {}
impl Sub for () {}

fn main() {
    let a: &dyn Sub = &();
    // `Super<fn(&'static ())>` and `Super<for<'a> fn(&'a ())>` have different vtables and we need to upcast to the latter!
    let b: &dyn Super<for<'a> fn(&'a ())> = a;
    b.call();
}
```

(or alternatively just do `vtable_principal == target_principal` (?))

Reply:
🤔 oh good point. that's kinda a different bug, though? but definitely one i could fix here. i think `==` is sufficient here, since we should be erasing relevant regions and anonymizing the binders... lemme think about it.

Reply:
(Anyways, for the record I believe this leak check bug was introduced with #131042)